python/pyiceberg/expressions/base.py

```diff
@@ -342,151 +313,197 @@ def __str__(self) -> str:
 
+    def bind(self, schema: Schema, case_sensitive: bool = True) -> BooleanExpression:
+        bound_term = self.term.bind(schema, case_sensitive)
+        return self.as_bound(bound_term, {lit.to(bound_term.ref().field.field_type) for lit in self.literals})
 
-class BoundNotNaN(BoundPredicate[T]):
-    def __invert__(self) -> BoundIsNaN:
-        return BoundIsNaN(self.term)
 
-    def _validate_literals(self):  # pylint: disable=W0238
-        if self.literals:
-            raise AttributeError("NotNaN is a unary predicate and takes no Literals.")
+@dataclass(frozen=True)
+class BoundSetPredicate(Bound[T], BooleanExpression, ABC):
+    term: BoundTerm[T]
+    literals: set[Literal[T]]
 
+    @abstractmethod
+    def __invert__(self) -> BooleanExpression:
+        ...
 
-class BoundIn(BoundPredicate[T]):
-    def _validate_literals(self):  # pylint: disable=W0238
-        if not self.literals:
-            raise AttributeError("BoundIn must contain at least 1 literal.")
 
-    def __invert__(self) -> BoundNotIn[T]:
-        return BoundNotIn(self.term, *self.literals)
+class BoundIn(BoundSetPredicate[T]):
+    def __new__(cls, term: BoundTerm[T], literals: set[Literal[T]]) -> BooleanExpression:
+        count = len(literals)
+        if count == 0:
+            return AlwaysFalse()
+        if count == 1:
+            return BoundEq(term, literals.pop())
+        else:
+            return super().__new__(cls)
 
+    def __invert__(self) -> BooleanExpression:
+        return BoundNotIn(self.term, self.literals)
 
-class In(UnboundPredicate[T]):
-    def _validate_literals(self):  # pylint: disable=W0238
-        if not self.literals:
-            raise AttributeError("In must contain at least 1 literal.")
 
-    def __invert__(self) -> NotIn[T]:
-        return NotIn(self.term, *self.literals)
+class BoundNotIn(BoundSetPredicate[T]):
+    def __new__(cls, term: BoundTerm[T], literals: set[Literal[T]]) -> BooleanExpression:
+        count = len(literals)
+        if count == 0:
+            return AlwaysTrue()
+        if count == 1:
+            return BoundNotEq(term, literals.pop())
+        else:
+            return super().__new__(cls)
 
-    def bind(self, schema: Schema, case_sensitive: bool) -> BoundIn[T]:
-        bound_ref = self.term.bind(schema, case_sensitive)
-        return BoundIn(bound_ref, *tuple(lit.to(bound_ref.field.field_type) for lit in self.literals))  # type: ignore
+    def __invert__(self) -> BooleanExpression:
+        return BoundIn(self.term, self.literals)
 
 
-class BoundNotIn(BoundPredicate[T]):
-    def _validate_literals(self):  # pylint: disable=W0238
-        if not self.literals:
-            raise AttributeError("BoundNotIn must contain at least 1 literal.")
+class In(SetPredicate[T]):
+    as_bound = BoundIn
 
-    def __invert__(self) -> BoundIn[T]:
-        return BoundIn(self.term, *self.literals)
+    def __new__(cls, term: UnboundTerm[T], literals: tuple[Literal[T], ...]) -> BooleanExpression:
```

Comment:
How about accumulating the positional arguments like:
```suggestion
    def __new__(cls, term: UnboundTerm[T], *literals: Literal[T]) -> BooleanExpression:
```

This way we can write:
```python
def test_in_to_eq():
    assert base.In(
        base.Reference("x"),
        literal(34.56),
        literal(19.25)
    ) == base.Eq(base.Reference("x"), literal(34.56))
```
instead of:
```python
def test_in_to_eq():
    assert base.In(
        base.Reference("x"),
        (literal(34.56), literal(19.25))
    ) == base.Eq(base.Reference("x"), literal(34.56))
```
I think this makes the API a bit friendlier and less verbose.

You can easily expand the arguments when you want to supply a list:
```python
class NotIn(SetPredicate[T]):
    as_bound = BoundNotIn
...
    def __invert__(self) -> BooleanExpression:
        return In(self.term, *self.literals)
```

Reply:
I wanted to do this, but I didn't know how to make it work with `@dataclass`. I looked it up and saw that you have to implement `__init__`. But maybe we don't want to use `@dataclass` anyway? It seems to have strange mypy errors:

```
327: error: Only concrete class can be given where "Type[UnaryPredicate[T]]" is expected
```